core/include/gpio.h

```diff
@@ -23,16 +26,42 @@ enum gpio_interrupt {
 
 struct gpio_chip {
 	const struct gpio_ops *ops;
+	unsigned int gpio_base;
+	unsigned int ngpios;
+	vaddr_t base;
+
+	SLIST_ENTRY(gpio_chip) link;
 };
 
 struct gpio_ops {
-	enum gpio_dir (*get_direction)(unsigned int gpio_pin);
-	void (*set_direction)(unsigned int gpio_pin, enum gpio_dir direction);
-	enum gpio_level (*get_value)(unsigned int gpio_pin);
-	void (*set_value)(unsigned int gpio_pin, enum gpio_level value);
-	enum gpio_interrupt (*get_interrupt)(unsigned int gpio_pin);
-	void (*set_interrupt)(unsigned int gpio_pin,
-		enum gpio_interrupt ena_dis);
+	enum gpio_dir (*get_direction)(struct gpio_chip *gc, unsigned int pin);
+	void (*set_direction)(struct gpio_chip *gc, unsigned int pin,
+			      enum gpio_dir direction);
+	enum gpio_level (*get_value)(struct gpio_chip *gc, unsigned int pin);
+	void (*set_value)(struct gpio_chip *gc, unsigned int pin,
+			  enum gpio_level value);
+	enum gpio_interrupt (*get_interrupt)(struct gpio_chip *gc,
+					     unsigned int pin);
+	void (*set_interrupt)(struct gpio_chip *gc, unsigned int pin,
+			      enum gpio_interrupt ena_dis);
+};
+
+struct gpio_desc {
+	unsigned int pin;
+	struct gpio_chip *gc;
+	void *dev;
```

Comment:
looks line `dev` is not used. remove if so.
actually, I think the whole structure could be remove and `struct gpio_chip *gc` moved to the gpio chip structure . thoughts?

Reply:
gd->dev  is currently used only to check if its owned or free. gd->client_data or  gd->owner 
or something else will be more appropriate ?
 regarding the gpio_desc structure , as I mentioned in previous comment reply I dont find easy alternate to have all functionality without maintaining these two lists. gclist , gdlist.  Let me know
if I can do something else.

1. gclist :  list of gpio_chips . which can be one node or more in a platform . It contains baseaddress, ops, interruptid  etc for the controller. Like you cited some platforms can derive
most of this information from the pin number. Partially agree. But still you will have overhead to
do the derivation on each ops. Instead this can be simpler.
2. gdlist: for each used pins in a platform . it describes the pin . its controller, its consumer, registered interrupt handler by consumer and may be more.

Reply:
> > 1. gdlist: for each used pins in a platform . it describes the pin . its controller, its consumer, registered interrupt handler by consumer and may be more.
> 
> Why registering the interrupt handler from the GPIO descriptor? GPIO interrupts are interrupt. Are you proposing a GPIO event notification framework, aside from the interrupt framework `kernel/interrupt.h`?

Most GPIO controllers will have combined interrupt for a group of GPIOS , may be per bank/ per chip.  So some sort of event notification framework will simplify future drivers. I have not completed anything on that. I will post a draft patch on top of this.